### DISCUSSION TITLE
webui.bat does nothing at all

### DISCUSSION THREAD

**Discussion by unalignedcoder** [vladmandic/sdnext#2433] [General]
On Windows 10, Python ~~3~~3.11.6, NVIDIA 3060 card, via pyenv

When launching `.\webui.bat --debug `nothing seem to happen. I am returned to the prompt. Both in PowerShell and Commandline.
I have followed instructions, plus `pip install -r requirements.txt` (probably unnecessary, but to try something).

Auto1111 and EasyDiffusion work on this system... To be noted that I run auto1111 through an extra `run.bat` file, following the method described [here](https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Install-and-Run-on-NVidia-GPUs). Maybe something similar is needed for SD.Next?

**Comment by brknsoul** [comment]
> Python 33.11.6

Assuming you've installed Python 3.11.6, and you're not from the future.

SD.Next Requires a Python version between 3.10.6 and 3.10.12.

Installing SD.Next is rather easy,
Install Python 3.10
Install Git
Open a command prompt where you'd like SD.Next to be installed, then;
`git clone https://github.com/vladmandic/automatic`
then run `webui.bat` inside the `automatic` folder.

**Reply by unalignedcoder** [reply]
Sorry for my typo... nevertheless... It seems strange it would fail silently because of this...

As I said, I've followed all instructions, except for the Python 3.10 thing which somehow escaped me.

At the moment I am running `py launch.py` directly, on Python 3.11, and it hasn't failed yet (still slowly reinstalling dependencies).

First few lines:
```
Starting SD.Next
INFO     Python 3.11.6 on Windows
INFO     Version: app=sd.next updated=2023-10-31 hash=2336ffc0
                         url=https://github.com/vladmandic/automatic/tree/master
INFO     Platform: arch=AMD64 cpu=Intel64 Family 6 Model 165 Stepping 5, GenuineIntel system=Windows
                         release=Windows-10-10.0.19045-SP0 python=3.11.6
INFO     nVidia CUDA toolkit detected: nvidia-smi present
```
You would think it would warn me of having the wrong Python

**Reply by brknsoul** [reply]
I believe SD.Next supports 3.11, some underlying modules and extensions might not.

**Reply by unalignedcoder** [reply]
If I manage to install and run this, will report.

For what is worth, the installation instructions say nothing about a required Python version:

![image](https://github.com/vladmandic/automatic/assets/16850566/2dc0cfbd-af08-4484-a234-35f490f9ac7d)

**Reply by unalignedcoder** [reply]
oh, wait, I see now.. that link is for Python 3.10! But, if the user has python already installed, they will not click on the link. Or they may think it's an old link.

**Reply by unalignedcoder** [reply]
Installed 3.10.11:

`pyenv install 3.10.11`
`pyenv local 3.10.11`
`./webui.bat --debug`

still fails silently (returns to prompt after a second).

Running `py launch.py` directly failed because of no checkpoints:
```
ERROR    Cannot generate without a checkpoint
INFO     Set system paths to use existing folders in a different location
INFO     Or use --ckpt <path-to-checkpoint> to force using existing checkpoint
INFO     Startup time: 83.30s { torch=22.66s gradio=2.77s diffusers=0.10s libraries=6.98s
                         extensions=38.80s face-restore=1.34s upscalers=0.50s extra-networks=0.12s
                         ui-extra-networks=3.22s ui-txt2img=0.06s ui-img2img=0.09s ui-settings=3.69s
                         ui-extensions=1.40s ui-defaults=0.06s launch=0.36s api=0.47s app-started=0.24s
                         checkpoint=0.32s }
```
after the above message, the script hangs.

**Reply by brknsoul** [reply]
Why are you using pyenv? webui will create its own environment.

Also, that isn't hanging. Either navigate to 127.0.0.1:7860 or use `--autolaunch`

**Reply by unalignedcoder** [reply]
Well, it seems to be working now, via `py launch.py --autolaunch`. Thanks for the advice. Still not clear why webui.bat won't work and what is its purpose, I am sure there is one. (Creating the python env?)

**Comment by vladmandic** [comment]
@brknsoul sdnext supports python 3.10.x, 3.11.x and 3.12.x, but some packages may not exist on 3.12 - for example `torch-directml`. that would show up during install step, so its pretty visible. i haven't heard of issues with 3.11.

@unalignedcoder pretty much all i can say is wtf is going on:
- `pip install -r requirements.txt` is not mentioned, not required and NOT SUPPORTED  
  sdnext installer takes care of its requirements by itself.
- `pyenv` is not SUPPORTED if you're using `webui.bat`/`webui.sh`
  those startup scripts expect to work with just pure python and they create venv
- when you run `py launch.py`, it basically bypasses creation of venv and everything is installed in global namespace
  does it work? yup. but you're on your own  
  chances of that creating conflicts with other python apps on your system is >90%. the whole point of venv is  
  so sdnext can have its own requirements that are not messed with other apps.
- "script hangs" - wtf? its not hanging
  startup is done and ready.
- what else?  
  this is a massive list of self-inflicted problems

to go back to step one:
- have `python` (not `py`) available in your path
- run `webui` and it creates venv, installs requirements as needed and starts sdnext

**Reply by unalignedcoder** [reply]
Wait a second. I mean, thanks for your answer. True, I have pyenv (it's a requirement in another programI am running), but obviously I also have Python installed in my path and originally I did not try to use pyenv with this script at all. I only did after I was told the version was not supported. 

Initially, I followed instructions to the letter and just double-clicked on webui.bat and it died. Why, if I have Python 3.11.6 in my path?

**Reply by vladmandic** [reply]
If you double clicked on it, it said SOMETHING.
If your console auto-closed, that's up to your console. Open console and run it so it doesn't auto-close, don't just double click.
And if you reported what it said, we could have taken it from there.

**Reply by unalignedcoder** [reply]
That's the thing. As per my first post. I open it, and I am brought back to the prompt. No messages whatsoever. Yes I open the console first and then run the command ("double-click" here is a figure of speech).
I tried this in PowerShell and Command Prompt. Same result.

**Reply by brknsoul** [reply]
So you literally see;
```
X:\SomeFolder\> webui
X:\SomeFolder\> _
``` 

?
What's the contents of webui.bat?

**Reply by unalignedcoder** [reply]
YES.
I see this:

![image](https://github.com/vladmandic/automatic/assets/16850566/a18cea95-04ae-4902-b36f-a1954f345995)

Can make an animated gif if you need.

**Reply by unalignedcoder** [reply]
webui.bat content:
```

@echo off

if not defined PYTHON (set PYTHON=python)
if not defined VENV_DIR (set "VENV_DIR=%~dp0%venv")
set ERROR_REPORTING=FALSE
mkdir tmp 2>NUL

%PYTHON% -c "" >tmp/stdout.txt 2>tmp/stderr.txt
if %ERRORLEVEL% == 0 goto :check_pip
echo Cannot launch python
goto :show_stdout_stderr

:check_pip
%PYTHON% -mpip --help >tmp/stdout.txt 2>tmp/stderr.txt
if %ERRORLEVEL% == 0 goto :start_venv
if "%PIP_INSTALLER_LOCATION%" == "" goto :show_stdout_stderr
%PYTHON% "%PIP_INSTALLER_LOCATION%" >tmp/stdout.txt 2>tmp/stderr.txt
if %ERRORLEVEL% == 0 goto :start_venv
echo Cannot install pip
goto :show_stdout_stderr

:start_venv
if ["%VENV_DIR%"] == ["-"] goto :skip_venv
if ["%SKIP_VENV%"] == ["1"] goto :skip_venv

dir "%VENV_DIR%\Scripts\Python.exe" >tmp/stdout.txt 2>tmp/stderr.txt
if %ERRORLEVEL% == 0 goto :activate_venv

for /f "delims=" %%i in ('CALL %PYTHON% -c "import sys; print(sys.executable)"') do set PYTHON_FULLNAME="%%i"
echo Using python: %PYTHON_FULLNAME%
echo Creating VENV: %VENV_DIR%
%PYTHON_FULLNAME% -m venv "%VENV_DIR%" >tmp/stdout.txt 2>tmp/stderr.txt
if %ERRORLEVEL% == 0 goto :activate_venv
echo Failed creating VENV: "%VENV_DIR%"
goto :show_stdout_stderr

:activate_venv
set PYTHON="%VENV_DIR%\Scripts\Python.exe"
echo Using VENV: %VENV_DIR%

:skip_venv
if [%ACCELERATE%] == ["True"] goto :accelerate
goto :launch

:accelerate
set ACCELERATE="%VENV_DIR%\Scripts\accelerate.exe"
if EXIST %ACCELERATE% goto :accelerate_launch

:launch
%PYTHON% launch.py %*
pause
exit /b

:accelerate_launch
echo Using accelerate
%ACCELERATE% launch --num_cpu_threads_per_process=6 launch.py %*
pause
exit /b

:show_stdout_stderr

echo.
echo exit code: %errorlevel%

for /f %%i in ("tmp\stdout.txt") do set size=%%~zi
if %size% equ 0 goto :show_stderr
echo.
echo stdout:
type tmp\stdout.txt

:show_stderr
for /f %%i in ("tmp\stderr.txt") do set size=%%~zi
if %size% equ 0 goto :show_stderr
echo.
echo stderr:
type tmp\stderr.txt

:endofscript

echo.
echo Launch Failed
pause
```

**Reply by brknsoul** [reply]
you're in Windows, try it without the `.\`

**Reply by unalignedcoder** [reply]
nope. Windows 10 does not allow launching without the .dot unless the script is in path. At least, in PowerShell. I get some error message otherwise:

![image](https://github.com/vladmandic/automatic/assets/16850566/7e47d35c-3383-4edf-909d-7e99f1a422a3)

In command line, admittedly, I don't need the dot, but I get the same result.

![image](https://github.com/vladmandic/automatic/assets/16850566/77c0462c-fc9e-46e0-953a-0092bb6de293)

**Reply by unalignedcoder** [reply]
I guess the script crashes? Perhaps I should look in the python logs?

**Reply by brknsoul** [reply]
What does `python --version` do?

**Reply by unalignedcoder** [reply]
`Python 3.11.6`
tmp/stderr and tmp/stout are empty btw. No logs.

**Reply by brknsoul** [reply]
could try `set ERROR_REPORTING=TRUE` in the bat file

All I can think is perhaps Defender or AV is blocking bat files 🤨

**Reply by unalignedcoder** [reply]
Tried, no difference. Pretty sure the script crashes early on. 

Defender is completely disabled, besides I have EasyDiffusion and Auto1111 .bat scripts running without any problem. 

I guess i'll try to completely remove pyenv and see if that's conflicting somehow with venv. Hate to do it because I have other projects requiring it.

**Reply by unalignedcoder** [reply]
YES. Removing pyenv from Windows and restarting allowed the script to run. It is now downloading torch.

So, a requirement for this script is that pyenv should not be installed at all?

**Reply by vladmandic** [reply]
requirement is that python is actual python. it doesn't know or care about pyenv.  
my guess is that pyenv is somehow managing what `python` means and its not usable by script.
you can do `set PYTHON="path_to_your_python_exe"` before running script and it will use that.

**Reply by Tillerz** [reply]
Add a 'pause' command to your startup batch script to see if the scripts throw any errors before exiting?

**Comment by gohan2091** [comment]
I'm running Python 3.10.11. Is there any benefits to installing a newer version such as Python 3.11.6?

**Reply by vladmandic** [reply]
No, not really.

**Comment by uzuuzu20002000** [comment]
I am always grateful. 

Not sure if it is the same, but it stopped starting with an error.
I have tried a clean install with the same result.
Is there a problem with my environment?

21:05:14-490280 INFO     Python 3.10.13 on Windows
21:05:15-252297 INFO     Version: app=sd.next updated=2023-11-13 hash=f1862579

21:05:57-058521 INFO     Load UI theme: name="black-teal" style=Auto base=style.css
╭─────────────────────────────────────────────────────────── Traceback (most recent call last) ────────────────────────────────────────────────────────────╮
│ C:\Users\uzuuz\AppData\Local\Programs\Python\Python310\Lib\site-packages\diffusers\utils\import_utils.py:710 in _get_module                              │
│                                                                                                                                                          │
│   709         try:                                                                                                                                       │
│ ❱ 710             return importlib.import_module("." + module_name, self.__name__)                                                                       │
│   711         except Exception as e:                                                                                                                     │
│                                                                                                                                                          │
│ C:\Users\uzuuz\AppData\Local\Programs\Python\Python310\lib\importlib\__init__.py:126 in import_module                                                    │
│                                                                                                                                                          │
│   125             level += 1                                                                                                                             │
│ ❱ 126     return _bootstrap._gcd_import(name[level:], package, level)                                                                                    │
│   127                                                                                                                                                    │
│ in _gcd_import:1050                                                                                                                                      │
│ in _find_and_load:1027                                                                                                                                   │
│ in _find_and_load_unlocked:1006                                                                                                                          │
│ in _load_unlocked:688                                                                                                                                    │
│ in exec_module:883                                                                                                                                       │
│ in _call_with_frames_removed:241                                                                                                                         │
│                                                                                                                                                          │
│ C:\Users\uzuuz\AppData\Local\Programs\Python\Python310\Lib\site-packages\diffusers\pipelines\stable_diffusion_xl\pipeline_stable_diffusion_xl.py:55 in   │
│ <module>                                                                                                                                                 │
│                                                                                                                                                          │
│     54 if is_torch_xla_available():                                                                                                                      │
│ ❱   55     import torch_xla.core.xla_model as xm                                                                                                         │
│     56                                                                                                                                                   │
│                                                                                                                                                          │
│ C:\Users\uzuuz\AppData\Local\Programs\Python\Python310\Lib\site-packages\torch_xla\__init__.py:1 in <module>                                             │
│                                                                                                                                                          │
│ ❱ 1 raise ImportError("WRONG PACKAGE. Please install the package from Neuron Repository - pi                                                             │
│   2                                                                                                                                                      │
╰──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
ImportError: WRONG PACKAGE. Please install the package from Neuron Repository - pip.repos.neuron.amazonaws.com

The above exception was the direct cause of the following exception:

╭─────────────────────────────────────────────────────────── Traceback (most recent call last) ────────────────────────────────────────────────────────────╮
│ I:\automatic\launch.py:225 in <module>                                                                                                                   │
│                                                                                                                                                          │
│   224                                                                                                                                                    │
│ ❱ 225     uv, instance = start_server(immediate=True, server=None)                                                                                       │
│   226     while True:                                                                                                                                    │
│                                                                                                                                                          │
│ I:\automatic\launch.py:162 in start_server                                                                                                               │
│                                                                                                                                                          │
│   161         else:                                                                                                                                      │
│ ❱ 162             uvicorn = server.webui(restart=not immediate)                                                                                          │
│   163     if args.profile:                                                                                                                               │
│                                                                                                                                                          │
│ I:\automatic\webui.py:311 in webui                                                                                                                       │
│                                                                                                                                                          │
│   310     start_common()                                                                                                                                 │
│ ❱ 311     start_ui()                                                                                                                                     │
│   312     modules.sd_models.write_metadata()                                                                                                             │
│                                                                                                                                                          │
│ I:\automatic\webui.py:232 in start_ui                                                                                                                    │
│                                                                                                                                                          │
│   231     timer.startup.record("before-ui")                                                                                                              │
│ ❱ 232     shared.demo = modules.ui.create_ui(timer.startup)                                                                                              │
│   233     timer.startup.record("ui")                                                                                                                     │
│                                                                                                                                                          │
│ I:\automatic\modules\ui.py:1071 in create_ui                                                                                                             │
│                                                                                                                                                          │
│   1070                         else:                                                                                                                     │
│ ❱ 1071                             component = create_setting_component(k)                                                                               │
│   1072                             component_dict[k] = component                                                                                         │
│                                                                                                                                                          │
│                                                                 ... 2 frames hidden ...                                                                  │
│                                                                                                                                                          │
│ I:\automatic\modules\shared_items.py:36 in get_pipelines                                                                                                 │
│                                                                                                                                                          │
│   35         'Stable Diffusion Upscale': getattr(diffusers, 'StableDiffusionUpscalePipeline',                                                            │
│ ❱ 36         'Stable Diffusion XL': getattr(diffusers, 'StableDiffusionXLPipeline', None),                                                               │
│   37         'Stable Diffusion XL Img2Img': getattr(diffusers, 'StableDiffusionXLImg2ImgPipel                                                            │
│                                                                                                                                                          │
│ C:\Users\uzuuz\AppData\Local\Programs\Python\Python310\Lib\site-packages\diffusers\utils\import_utils.py:701 in __getattr__                              │
│                                                                                                                                                          │
│   700             module = self._get_module(self._class_to_module[name])                                                                                 │
│ ❱ 701             value = getattr(module, name)                                                                                                          │
│   702         else:                                                                                                                                      │
│                                                                                                                                                          │
│ C:\Users\uzuuz\AppData\Local\Programs\Python\Python310\Lib\site-packages\diffusers\utils\import_utils.py:701 in __getattr__                              │
│                                                                                                                                                          │
│   700             module = self._get_module(self._class_to_module[name])                                                                                 │
│ ❱ 701             value = getattr(module, name)                                                                                                          │
│   702         else:                                                                                                                                      │
│                                                                                                                                                          │
│ C:\Users\uzuuz\AppData\Local\Programs\Python\Python310\Lib\site-packages\diffusers\utils\import_utils.py:700 in __getattr__                              │
│                                                                                                                                                          │
│   699         elif name in self._class_to_module.keys():                                                                                                 │
│ ❱ 700             module = self._get_module(self._class_to_module[name])                                                                                 │
│   701             value = getattr(module, name)                                                                                                          │
│                                                                                                                                                          │
│ C:\Users\uzuuz\AppData\Local\Programs\Python\Python310\Lib\site-packages\diffusers\utils\import_utils.py:712 in _get_module                              │
│                                                                                                                                                          │
│   711         except Exception as e:                                                                                                                     │
│ ❱ 712             raise RuntimeError(                                                                                                                    │
│   713                 f"Failed to import {self.__name__}.{module_name} because of the followin                                                           │
╰──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
RuntimeError: Failed to import diffusers.pipelines.stable_diffusion_xl.pipeline_stable_diffusion_xl because of the following error (look up to see its
traceback):
WRONG PACKAGE. Please install the package from Neuron Repository - pip.repos.neuron.amazonaws.com
